### PR TITLE
Nullable limit and before parameters in pagination, more precise Position type, multiple fixes

### PR DESCRIPTION
**Pull request by Lukellmann**
# Changes

## `Position`
- change to `sealed interface`
- introduce additional `Position.BeforeOrAfter` type to restrict position parameters in services
- type change `Position?` -> `Position.After?` in `ChannelService.getReactions()` and `GuildService.getGuildMembers()`
- type change `Position?` -> `Position.BeforeOrAfter?` in `UserService.getCurrentUserGuilds()`

## `paginate()` and `paginateByDate()`
- more specific types (e. g. `Position` -> `Direction : Position.BeforeOrAfter`)
- `size` will now always stay `batchSize` -> hypothetical page that is longer than `batchSize` will not lead to unwanted behavior in size comparison with next page
- changed order: `break` on smaller batch size before selecting next direction/timestamp

## nullable `limit` parameter for
- `AuditLogGetRequestBuilder`, `AuditLogGetRequest`
- `AuditLogService.getAuditLogs()`
- `getMessages[Before|After]()` in `MessageChannelBehavior`, `EntitySupplier`
- `getPublicArchivedThreads()` in `ThreadParentChannelBehavior`, `EntitySupplier`
- `get[Joined]PrivateArchivedThreads()` in `PrivateThreadParentChannelBehavior`, `EntitySupplier`
- `getGuildMembers()` in `EntitySupplier`, `GuildService`
- `getCurrentUserGuilds()` in `EntitySupplier`, `UserService`
- `getMessages()`, `getReactions()` in `ChannelService`

## nullable `before` parameter for
- `getPublicArchivedThreads()` in `ThreadParentChannelBehavior`, `EntitySupplier`
- `get[Joined]PrivateArchivedThreads()` in `PrivateThreadParentChannelBehavior`, `EntitySupplier`

## `CacheEntitySupplier`
- fix `getMessagesBefore()` behaving the same as `getMessagesAfter()` -> replace `idGt` with `lt`
- also emit message with `messageId` in `getMessagesAround()`
- fix flow from `getGuildMemebers()` not being limited
- check `limit` parameter in `get[...]ArchivedThreads()`

## `RestEntitySupplier`
- increase max pagination `batchSize` for `guilds` and `getCurrentUserGuilds()` to 200, see https://discord.com/developers/docs/resources/user#get-current-user-guilds and https://github.com/discord/discord-api-docs/commit/daa13f60634396296092da02bb8acd8f3360e03b
- higher-order `limitedPagination()` function that checks `limit`, calculates `batchSize`, creates the pagination flow and limits it
- check `limit` parameter in `getMessagesAround()`
- fix not using `limit` parameter in `getMessagesAround()`
- lazy `map` on flow instead of pagination batch directly to potentially safe requests made on `guild.getGuild()` in `getCurrentUserGuilds()`
- fix `getAuditLogEntries()` not using optimized `batchSize` and not adjusting `request.limit`
- fix not using `Snowflake`/`Instant` from pagination in  `get[...]ArchivedThreads()`

## `MessageChannelBehavior.getMessagesAround()`
- fix not using `limit` parameter

## `GuildBehavior`
- removed unnecessary `suspend` modifier on `getMembers()`

## `StoreEntitySupplier`
- deprecate constructor with `kord` parameter, parameter was unused

## Documentation
- remove no longer applicable warning in `GuildBehavior.getMembers()` doc
- add note about flow additionally containing message itself for `getMessagesAround()` in `MessageChannelBehavior`, `EntitySupplier`
- document different usages of `AuditLogGetRequestBuilder.limit`
- ...

### CODE REVIEW

File: core/src/main/kotlin/behavior/GuildBehavior.kt [hunk 1]
```diff
@@ -350,11 +350,9 @@ public interface GuildBehavior : KordEntity, Strategizable {
...
      * The returned flow is lazily executed, any [RequestException] will be thrown on
      * [terminal operators](https://kotlinlang.org/docs/reference/coroutines/flow.html#terminal-flow-operators) instead.
-     *
-     * This function is not part of the officially documented Discord API and may be removed/altered/stop working in the future.
      */
     @KordExperimental
-    public suspend fun getMembers(query: String, limit: Int = 1000): Flow<Member> = flow {
+    public fun getMembers(query: String, limit: Int = 1000): Flow<Member> = flow {
         kord.rest.guild.getGuildMembers(id, query, limit).forEach {
             emit(
```
Comment (HopeBaron) [inline core/src/main/kotlin/behavior/GuildBehavior.kt:357]: This sounds out of place, would be nice if indented properly

File: core/src/main/kotlin/supplier/StoreEntitySupplier.kt [hunk 2]
```diff
@@ -69,11 +75,11 @@ public class StoreEntitySupplier(
         return storeAndReturn(supplier.getMessageOrNull(channelId, messageId)) { it.data }
     }
 
-    override fun getMessagesAfter(messageId: Snowflake, channelId: Snowflake, limit: Int): Flow<Message> {
+    override fun getMessagesAfter(messageId: Snowflake, channelId: Snowflake, limit: Int?): Flow<Message> {
```
[the remark this answers is not in the document]
Comment (Lukellmann) [reply core/src/main/kotlin/supplier/StoreEntitySupplier.kt:78]: If you mean `limit` no - overrides can't have default parameters, it defaults to null in `EntitySupplier` anyway.

File: core/src/main/kotlin/behavior/GuildBehavior.kt (hunk 1 above)
Comment (Lukellmann) [reply core/src/main/kotlin/behavior/GuildBehavior.kt:357]: Not sure what you mean, every new level is indented 4 more spaces.

File: core/src/main/kotlin/supplier/CacheEntitySupplier.kt [hunk 3]
```diff
@@ -157,27 +147,28 @@ public class CacheEntitySupplier(private val kord: Kord) : EntitySupplier {
...
+        }.asFlow().map { Message(it, kord) }.limit(limit)
     }
 
-    override fun getMessagesBefore(messageId: Snowflake, channelId: Snowflake, limit: Int): Flow<Message> {
-        require(limit > 0) { "At least 1 item should be requested, but got $limit." }
+    override fun getMessagesBefore(messageId: Snowflake, channelId: Snowflake, limit: Int?): Flow<Message> {
+        checkLimit(limit)
         return cache.query<MessageData> {
             idEq(MessageData::channelId, channelId)
-            idGt(MessageData::id, messageId)
```
Comment (HopeBaron) [inline core/src/main/kotlin/supplier/CacheEntitySupplier.kt:172]: make an equvilant function for Lt, it'd be better to have common conversion OVA.

**Pull request merged**